File: app/src/main/java/com/app/ralaunch/controls/editor/manager/ControlDataSyncManager.java
package com.app.ralaunch.controls.editor.manager;

import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.app.ralaunch.controls.ControlData;
import com.app.ralaunch.controls.ControlLayout;
import com.app.ralaunch.controls.ControlView;

/**
 * 控件数据同步管理器
 * 统一管理 ControlData 到 ControlView 的同步逻辑
 */
public class ControlDataSyncManager {

    /**
     * 同步控件数据到视图
     * @param layout 控件布局
     * @param controlData 要同步的控件数据
     * @return 是否成功同步
     */
    public static boolean syncControlDataToView(ControlLayout layout, ControlData controlData) {
        if (layout == null || controlData == null) {
            return false;
        }

        for (int i = 0; i < layout.getChildCount(); i++) {
            View child = layout.getChildAt(i);
            if (child instanceof ControlView) {
                ControlView controlView = (ControlView) child;
                ControlData viewData = controlView.getData();
                
                // 优先使用引用比较，确保匹配到正确的控件
                // 只有在引用不同时才进行名称比较（作为后备方案）
                boolean isMatch = false;
                if (viewData == controlData) {
                    // 引用匹配 - 最可靠的方式
                    isMatch = true;
                } else if (viewData != null && controlData != null) {
                    // 引用不同，尝试名称匹配（仅作为后备，可能存在名称重复的情况）
                    // 但这里我们更倾向于使用引用匹配，所以只在确实需要时才使用名称匹配
                    // 实际上，我们应该避免使用名称匹配，因为名称可能被用户修改
                    // 为了安全，我们只在引用匹配时才更新
                    isMatch = false; // 禁用名称匹配，避免匹配到错误的控件
                }
                
                if (isMatch) {
                    // 更新布局参数
                    ViewGroup.LayoutParams layoutParams = child.getLayoutParams();
                    if (layoutParams instanceof FrameLayout.LayoutParams) {
                        FrameLayout.LayoutParams frameParams = (FrameLayout.LayoutParams) layoutParams;
                        frameParams.width = (int) controlData.width;
                        frameParams.height = (int) controlData.height;
                        frameParams.leftMargin = (int) controlData.x;
                        frameParams.topMargin = (int) controlData.y;
                        child.setLayoutParams(frameParams);
                    }
                    
                    // 更新视觉属性
                    // 对于摇杆，不设置 View 级别的 alpha，因为它们有独立的透明度控制
                    // View 级别的 alpha 会影响所有绘制内容，导致透明度叠加
                    if (controlData.type == ControlData.TYPE_JOYSTICK) {
                        // 摇杆：View alpha 始终为 1.0，让内部的 Paint alpha 独立控制
                        child.setAlpha(1.0f);
                    } else {
                        // 按钮和文本：可以使用 View 级别的 alpha
                        child.setAlpha(controlData.opacity);
                    }
                    child.setVisibility(controlData.visible ? View.VISIBLE : View.INVISIBLE);
                    
                    // 同步所有字段到原始数据对象，确保数据一致性
                    if (viewData != null) {
                        syncDataFields(viewData, controlData);
                    }
                    
                    // 刷新控件绘制
                    controlView.updateData(controlData);
                    child.invalidate();
                    
                    return true;
                }
            }
        }
        
        return false;
    }

    /**
     * 同步数据字段
     * 将源数据的所有字段同步到目标数据对象
     * 注意：如果 target 和 source 是同一个对象，则不需要同步（避免不必要的操作）
     */
    private static void syncDataFields(ControlData target, ControlData source) {
        if (target == null || source == null) {
            return;
        }
        
        // 如果 target 和 source 是同一个对象，不需要同步
        if (target == source) {
            return;
        }
        
        // 同步所有关键字段
        target.type = source.type;
        target.buttonMode = source.buttonMode;
        target.shape = source.shape;
        target.keycode = source.keycode;
        target.isToggle = source.isToggle;
        target.opacity = source.opacity;
        target.borderOpacity = source.borderOpacity;
        target.textOpacity = source.textOpacity;
        target.visible = source.visible;
        target.bgColor = source.bgColor;
        target.strokeColor = source.strokeColor;
        target.strokeWidth = source.strokeWidth;
        target.cornerRadius = source.cornerRadius;
        target.name = source.name;
        // 位置信息必须同步，因为用户可能拖动过控件
        target.x = source.x;
        target.y = source.y;
        target.width = source.width;
        target.height = source.height;
        
        // 同步摇杆特有属性
        if (source.type == ControlData.TYPE_JOYSTICK) {
            target.stickOpacity = source.stickOpacity;
            target.stickKnobSize = source.stickKnobSize;
            if (source.joystickKeys != null) {
                target.joystickKeys = source.joystickKeys.clone();
            }
            // 深拷贝统一组合键数组
            if (source.joystickComboKeys != null) {
                target.joystickComboKeys = source.joystickComboKeys.clone();
            } else {
                target.joystickComboKeys = new int[0];
            }
            target.joystickMode = source.joystickMode;
            target.xboxUseRightStick = source.xboxUseRightStick;
        }
        
        // 同步文本控件特有属性
        if (source.type == ControlData.TYPE_TEXT) {
            // 对于文本控件，允许 displayText 为空字符串
            // 如果 source.displayText 为 null，使用空字符串而不是 source.name
            target.displayText = source.displayText != null ? source.displayText : "";
        }
    }
}

File: app/src/main/java/com/app/ralaunch/controls/editors/managers/ControlDataSyncManager.kt
package com.app.ralaunch.controls.editors.managers

import android.util.Log
import android.view.View
import android.widget.FrameLayout
import com.app.ralaunch.controls.data.ControlData
import com.app.ralaunch.controls.views.ControlLayout
import com.app.ralaunch.controls.views.ControlView

/**
 * 控件数据同步管理器
 * 统一管理 ControlData 到 ControlView 的同步逻辑
 */
object ControlDataSyncManager {
    private const val TAG = "ControlDataSyncManager"
    
    /**
     * 同步控件数据到视图
     * @param layout 控件布局
     * @param controlData 要同步的控件数据
     * @return 是否成功同步
     */
    @JvmStatic
    fun syncControlDataToView(layout: ControlLayout?, controlData: ControlData?): Boolean {
        if (layout == null || controlData == null) {
            Log.w(TAG, "syncControlDataToView: layout=$layout, controlData=$controlData")
            return false
        }
        
        Log.d(TAG, "syncControlDataToView: looking for '${controlData.name}', childCount=${layout.childCount}")

        for (i in 0 until layout.childCount) {
            val child = layout.getChildAt(i)
            if (child is ControlView) {
                val viewData: ControlData? = child.controlData

                // 使用名称匹配，支持 Activity 重建后的反序列化对象
                val isMatch = viewData != null && (
                    viewData === controlData ||  // 引用相同（优先）
                    viewData.name == controlData.name  // 名称相同（备用）
                )
                
                Log.d(TAG, "  child[$i]: '${viewData?.name}', isMatch=$isMatch")

                if (isMatch) {
                    // Get screen dimensions from the layout
                    val screenWidth = layout.resources.displayMetrics.widthPixels
                    val screenHeight = layout.resources.displayMetrics.heightPixels

                    // Update layout parameters
                    // Convert fraction-based values to pixels
                    // Note: width and height are both relative to screen height
                    val layoutParams = child.layoutParams
                    if (layoutParams is FrameLayout.LayoutParams) {
                        layoutParams.width = (controlData.width * screenHeight).toInt()
                        layoutParams.height = (controlData.height * screenHeight).toInt()
                        layoutParams.leftMargin = (controlData.x * screenWidth).toInt()
                        layoutParams.topMargin = (controlData.y * screenHeight).toInt()
                        child.layoutParams = layoutParams
                    }

                    // 更新视觉属性
                    // 对于摇杆，不设置 View 级别的 alpha，因为它们有独立的透明度控制
                    if (controlData is ControlData.Joystick) {
                        // 摇杆：View alpha 始终为 1.0，让内部的 Paint alpha 独立控制
                        child.alpha = 1.0f
                    } else {
                        // 按钮和文本：可以使用 View 级别的 alpha
                        child.alpha = controlData.opacity
                    }
                    child.visibility = if (controlData.isVisible) View.VISIBLE else View.INVISIBLE

                    // 同步所有字段到原始数据对象，确保数据一致性
                    syncDataFields(viewData, controlData)
                    
                    // 验证纹理是否同步成功
                    if (viewData is ControlData.Button && controlData is ControlData.Button) {
                        Log.i(TAG, "Button texture synced: path='${viewData.texture.normal.path}', enabled=${viewData.texture.normal.enabled}")
                    }

                    // 重新设置 controlData 以触发 initPaints() 更新颜色/透明度
                    // 这是必须的，因为直接修改字段不会触发 Paint 重新初始化
                    child.controlData = viewData
                    
                    // 刷新控件绘制
                    child.invalidate()
                    
                    Log.i(TAG, "syncControlDataToView: SUCCESS for '${controlData.name}'")
                    return true
                }
            }
        }
        
        Log.w(TAG, "syncControlDataToView: FAILED - no matching control found for '${controlData.name}'")
        return false
    }

    /**
     * 同步数据字段
     * 将源数据的所有字段同步到目标数据对象
     * 注意：如果 target 和 source 是同一个对象，则不需要同步（避免不必要的操作）
     */
    private fun syncDataFields(target: ControlData, source: ControlData) {
        // 如果 target 和 source 是同一个对象，不需要同步
        if (target === source) {
            return
        }

        // 同步基础字段
        target.name = source.name
        target.x = source.x
        target.y = source.y
        target.width = source.width
        target.height = source.height
        target.rotation = source.rotation
        target.opacity = source.opacity
        target.borderOpacity = source.borderOpacity
        target.textOpacity = source.textOpacity
        target.bgColor = source.bgColor
        target.strokeColor = source.strokeColor
        target.strokeWidth = source.strokeWidth
        target.cornerRadius = source.cornerRadius
        target.isVisible = source.isVisible
        target.isPassThrough = source.isPassThrough

        // 同步特定类型的字段
        when {
            source is ControlData.Button && target is ControlData.Button -> {
                target.mode = source.mode
                target.keycode = source.keycode
                target.isToggle = source.isToggle
                target.shape = source.shape
                // 同步纹理配置
                target.texture = source.texture.copy()
            }
            source is ControlData.Joystick && target is ControlData.Joystick -> {
                target.stickKnobSize = source.stickKnobSize
                target.stickOpacity = source.stickOpacity
                target.joystickKeys = source.joystickKeys.clone()
                target.mode = source.mode
                target.isRightStick = source.isRightStick
                // 同步纹理配置
                target.texture = source.texture.copy()
            }
            source is ControlData.Text && target is ControlData.Text -> {
                target.displayText = source.displayText
                target.shape = source.shape
                // 同步纹理配置
                target.texture = source.texture.copy()
            }
            source is ControlData.TouchPad && target is ControlData.TouchPad -> {
                // 同步纹理配置
                target.texture = source.texture.copy()
            }
        }
    }
}

